File: bin/summary.R
#!/usr/bin/env Rscript
library(data.table)
library(stringr)
library(argparse)
library(dplyr)
library(R.utils)

parser <- ArgumentParser("Parameters for summarizing results")
parser$add_argument("--gene_demulti", help = "Folder containing output files of genetic demultiplexing pipeline", default = NULL)
parser$add_argument("--hash_demulti", help = "Folder containing output files of hashing demultiplexing pipeline", default = NULL)
args <- parser$parse_args()
assignment_gene <- list.files(args$gene_demulti, pattern = "_assignment_all.csv", full.names = TRUE)
assignment_hash <- list.files(args$hash_demulti, pattern = "_assignment_all.csv", full.names = TRUE)
assignment <- append(assignment_gene, assignment_hash)

ifelse(!dir.exists("summary"), dir.create("summary"), FALSE)

assignment_all <- lapply(assignment, function(x){
  assign <- fread(x, header = TRUE)
  assign
}) %>% Reduce(function(dtf1,dtf2) full_join(dtf1,dtf2,by="Barcode"), .)
assignment_all[assignment_all == "DBL"] <- "doublet"
assignment_all[assignment_all == "AMB"] <- "negative"
#write.table(assignment_all, "summary/assignment_all_genetic_and_hash.csv", row.names=F, quote=F, sep='\t')
write.table(assignment_all, "summary/assignment_all_genetic_and_hash.csv", row.names=F, quote=F)

classification_gene <- list.files(args$gene_demulti, pattern = "_classification_all.csv", full.names = TRUE)
classification_hash <- list.files(args$hash_demulti, pattern = "_classification_all.csv", full.names = TRUE)
classification <- append(classification_gene, classification_hash)
classification_all <- lapply(classification, function(x){
  classi <- fread(x, header = TRUE)
  classi
}) %>% Reduce(function(dtf1,dtf2) full_join(dtf1,dtf2,by="Barcode"), .)
#classification_all <- classification_all[, -c("count_sin", "count_dou", "count_neg")]
classification_all[classification_all == "SNG"] <- "singlet"
classification_all[classification_all == "DBL"] <- "doublet"
classification_all[classification_all == "AMB"] <- "negative"
#write.table(classification_all, "summary/classification_all_genetic_and_hash.csv", row.names=F, quote=F, sep='\t')
write.table(classification_all, "summary/classification_all_genetic_and_hash.csv", row.names=F, quote=F)

File: bin/summary.py
#!/usr/bin/env python
import argparse
import pandas as pd
from pathlib import Path

parser = argparse.ArgumentParser(description="Parameters for summarizing results")
parser.add_argument(
    "--gene_demulti",
    help="Folder containing output files of genetic demultiplexing pipeline",
    default=None,
)
parser.add_argument(
    "--hash_demulti",
    help="Folder containing output files of hashing demultiplexing pipeline",
    default=None,
)
args = parser.parse_args()


def merge_dataframes(dataframes: list[pd.DataFrame]) -> pd.DataFrame:
    merged_df = pd.DataFrame()
    for df in dataframes:
        if merged_df.empty:
            merged_df = df
        else:
            merged_df = pd.merge(merged_df, df, on="Barcode", how="outer")
    return merged_df


def find_first_file(directory: Path, suffix: str) -> Path:
    return [
        file
        for file in directory.iterdir()
        if file.name.endswith(suffix) and not file.name.startswith(".")
    ][0]


def process_file_pair(
    gene_dir: Path,
    hash_dir: Path,
    suffix: str,
    output_path: Path,
    replacements: dict[str, str],
) -> None:
    gene_file = find_first_file(gene_dir, suffix)
    hash_file = find_first_file(hash_dir, suffix)

    gene_df = pd.read_csv(gene_file, dtype=str)
    hash_df = pd.read_csv(hash_file, dtype=str)

    merged_df = merge_dataframes([gene_df, hash_df])
    merged_df = merged_df.replace(replacements)
    merged_df.to_csv(output_path, index=False, sep="\t")


if __name__ == "__main__":
    summary_dir = Path("summary")
    summary_dir.mkdir(exist_ok=True)

    gene_dir = Path(args.gene_demulti)
    hash_dir = Path(args.hash_demulti)

    process_file_pair(
        gene_dir,
        hash_dir,
        "_assignment_all.csv",
        summary_dir / "assignment_all_genetic_and_hash.csv",
        {"DBL": "doublet", "AMB": "negative"},
    )

    process_file_pair(
        gene_dir,
        hash_dir,
        "_classification_all.csv",
        summary_dir / "classification_all_genetic_and_hash.csv",
        {"SNG": "singlet", "DBL": "doublet", "AMB": "negative"},
    )
